Devel/analysis v3: latest changes

stephlazar opened the pull request:
Latest changes in analysis_v3:

- adds helper_functions.py/get_instr_param_from_hdf_file
- minor improvements in helper_functions.py/read_analysis_file
- ProcessingPipeline.resolve_keys_in(): just append keys_in that do not satisfy any of the special conditions
- process_tomography_analysis: clean-up and extended way for how ideal process is specified

Commit:
ProcessingPipeline.resolve_keys_in: add line to leave keys_in unmodified if none of the existing conditions are met

Commit:
ana_v3/read_analysis_file: fixes bug in exception and removes underscore from default file_id name

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
ana_v3/state_tomography_analysis: bugfix: meas_obj_names passed both as argument and keyword argument

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
helper_functions.py: add get_instr_param_from_hdf_file

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
analysis_v3/tomography_analysis.py: minor changes in obtaining Uideal in process_tomography_analysis

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

chellings commented on pycqed/analysis_v3/helper_functions.py:
```diff
@@ -1023,16 +1051,20 @@ def read_analysis_file(timestamp=None, filepath=None, data_dict=None,
+            try:
+                ana_file.close()
+            except AttributeError:
+                pass
```
If you silently pass an error, it might be good to add a comment that explains why.

stephlazar replied:
added now

chellings commented on pycqed/analysis_v3/processing_pipeline.py:
```diff
@@ -516,6 +516,8 @@ def resolve_keys_in(self, keys_in, mobj_name, meas_obj_value_names_map,
                     raise ValueError('The first node in the pipeline cannot '
                                      'have "keys_in" = "previous".')
+            else:
+                keys_in += [keyi]
```
Am I getting it wrong or is this the same action as in the first branch of the if? If yes, the code could be simplified. It might also be useful to add a comment why this is the default behavior.

stephlazar replied:
Good catch! I now removed the first if and kept the else. Plus added a comment under the else

chellings commented on pycqed/analysis_v3/tomography_analysis.py:
```diff
@@ -929,18 +927,27 @@ def process_tomography_analysis(data_dict, gate_name='CZ', Uideal=None,
-                             f'Please provide the Uideal.')
+            raise ValueError(f'Unknown gate of interest {process_name}. '
+                             f'Please provide the process unitary, Uideal.')
+    print(Uideal)
```
debug output

stephlazar replied:
Removed. Thanks

chellings commented on pycqed/analysis_v3/tomography_analysis.py:
```diff
@@ -929,18 +927,27 @@ def process_tomography_analysis(data_dict, gate_name='CZ', Uideal=None,
             measured_error_{gate_name}.{estimation_type} for estimation_type
             in estimation_types.
+        -  process_name: name of the gate for which the error is estimated.
+         MUST CORRESPOND TO Uideal IF THE LATTER IS PROVIDED, since gate_name
```
If this is a requirement, it might make sense to
- add a check whether the requirement is fulfilled
- better explain in what way the correspondence is meant, e.g., where can I look up the names that are expected for certain unitaries. Or do you mean the following? "If both Uideal and process_name are provided, process_name will only be used in the key name for storing the results, and it is up to the user to ensure that process_name is a meaningful description of the unitary Uideal."  --- but this is not what the code currently implements.

Moreover, the reference to the parameter "gate_name" seems to be outdated.

stephlazar replied:
I updated the docstring now and also fixed a bug that I realized after reading your comment. Thanks

chellings reviewed (changes requested):
Thanks for the PR. Please check my comments on the tomo analysis. It looks like a bug to me. For the other modules, I only have minor comments.

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
helper_functions.py/read_analysis_file: adds comment explaining why AttributeError is silently passed

Commit:
ProcessingPipeline.resolve_keys_in: merges two conditional statements

Commit:
analysis_v3/process_tomography_analysis: fixes bug and docstring related to specifying process_name

- process_name was required but it wasn't always extracted and its necessity was not enforced
- docstring was outdated regarding process_name

Commit:
tomography_analysis.py: adds chi_ideal to the data dict in process_tomography_analysis

Commit:
data_processing.py/calculate_flat_multiqubit_shots: adds detailed comments explaining how the correlations are calculated.

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

chellings commented on pycqed/analysis_v3/tomography_analysis.py:
```diff
@@ -915,7 +913,14 @@ def process_tomography_analysis(data_dict, gate_name='CZ', Uideal=None,
+                to the process Uideal for meaningful storing names.
+                If Uideal is None, specifying process_name = 'CZ' or 'CNOT'
+                will create the corresponding Uideal. Other gates are not
+                recofnized yet.
```
Typo recofnized

stephlazar replied:
👍

chellings commented on pycqed/analysis_v3/tomography_analysis.py:
```diff
@@ -926,21 +931,31 @@ def process_tomography_analysis(data_dict, gate_name='CZ', Uideal=None,
         n_qubits = len(meas_obj_names)
+
+    process_name = hlp_mod.get_param('process_name', data_dict,
+                                     raise_error=True, **params)
```
A general comment: it might be a bit confusing that process_name is mandatory even though it is only in **params, but I guess it has to be in params to allow to pass it via the data_dict as an alternative. So this probably fine, but might be something that could be part of a potential future general discussion about structural questions of v3.

stephlazar replied:
Yes, this good point that I've also noticed but indeed would need a longer discussion relating to analysis_v3 structure.

chellings commented on pycqed/analysis_v3/tomography_analysis.py:
```diff
@@ -1063,16 +1066,16 @@ def process_tomography_analysis(data_dict, Uideal=None,
         # add gate error to data_dict
         hlp_mod.add_param(f'measured_error_{process_name}.{estimation_type}',
-                          1-np.real(qtp.process_fidelity(chi_qtp, Uideal)),
+                          1-np.real(qtp.process_fidelity(chi_qtp, chi_ideal)),
```
I do not understand this change and it does not become clear to me from the commit message, but I assume that you know what you are doing :-)

stephlazar replied:
It is just a renaming of the variable because in line 954 Uideal is converted to a chi matrix so to keep it still as Uideal was misleading I thought.

chellings reviewed (approved):
Thanks for addressing my comments. I approve with some minor comments.

Commit:
process_tomography_analysis: Fixes typo in docstring

stephlazar commented:
Branch intentionally not deleted because it is my main development branch.
